tests/opentracing_instrumentation/test_requests.py

```diff
@@ -0,0 +1,113 @@
+        response_handler_hook.called_with = None
+    else:
+        response_handler_hook = None
+
+    patcher.install_patches()
+    patcher.set_response_handler_hook(response_handler_hook)
+    try:
+        yield response_handler_hook
+    finally:
+        patcher.reset_patches()
+
+
+@pytest.fixture
+def tornado_app(request, _unused_port):
+
+    class Handler(tornado.web.RequestHandler):
+        def get(self):
+            app.headers = self.request.headers
+
+    app = tornado.web.Application([('/', Handler)])
+
+    def run_http_server():
+        io_loop = tornado.ioloop.IOLoop.current()
+        http_server = tornado.httpserver.HTTPServer(app)
+        http_server.add_socket(_unused_port[0])
+
+        def stop():
+            http_server.stop()
+            io_loop.stop()
+            thread.join()
+
+        # finalizer should be added before starting of the IO loop
+        request.addfinalizer(stop)
+
+        io_loop.start()
+
+    # running an http server in a separate thread in purpose
+    # to make it accessible for the requests from the current thread
+    thread = threading.Thread(target=run_http_server)
+    thread.start()
+
+    return app
+
+
+def _test_requests(url, root_span, tracer, response_handler_hook, get_headers):
+    if root_span:
+        root_span = tracer.start_span('root-span')
+    else:
+        root_span = None
+
+    with span_in_context(span=root_span):
+        response = requests.get(url)
+
+    assert len(tracer.recorder.get_spans()) == 1
+
+    span = tracer.recorder.get_spans()[0]
+    assert span.tags.get('span.kind') == 'client'
+    assert span.tags.get('http.url') == url
+
+    # verify trace-id was correctly injected into headers
+    headers = get_headers()
+    trace_id = headers.get('ot-tracer-traceid')
+    assert trace_id == '%x' % span.context.trace_id
+
+    if response_handler_hook:
+        assert response_handler_hook.called_with == (response, span)
+
+
+@pytest.mark.parametrize('scheme', ('http', 'https'))
+@pytest.mark.parametrize('root_span', (True, False))
+@pytest.mark.parametrize('hook', (True, False))
+@mock.patch('requests.adapters.HTTPAdapter.cert_verify')
+@mock.patch('requests.adapters.HTTPAdapter.get_connection')
+def test_requests_with_mock(get_connection_mock, cert_verify_mock,
+                            scheme, root_span, tracer,
+                            response_handler_hook):
+    def get_headers():
+        return get_connection_mock.return_value.urlopen.call_args[1]['headers']
+
+    url = scheme + '://example.com/'
+    _test_requests(url, root_span, tracer, response_handler_hook, get_headers)
+
+
+@pytest.mark.parametrize('root_span', (True, False))
+@pytest.mark.parametrize('hook', (True, False))
+def test_requests_with_tornado(base_url, tornado_app, root_span, tracer,
+                               response_handler_hook):
+    def get_headers():
+        # feel free to suggest a better way to expose headers
```

Comment:
What I typically do in these tests is this:
  * add instrumentation to the server to create the server span from inbound request 
  * have the server return the trace ID of that server span in the response
  * have the caller verify that the returned trace ID matches the trace ID of the root span.

Doing so does not expose the test to the knowledge of which headers are being used by the tracer, and per [rfc/trace_identifiers](https://github.com/opentracing/specification/blob/master/rfc/trace_identifiers.md) won't even have dependency on the implementation details of the mock tracer (accessing the trace ID).

Reply:
Thank you for the suggestion!
I used a bit simpler approach to avoid mocking of the request wrapper when using mock for requests.
Please review the changes and let me know whether it's fine enough.